app/controllers/offices_controller.rb

```diff
@@ -1,22 +1,41 @@
 class OfficesController < ApplicationController
   def index
+    # return on direct requests to the page
     return unless search
-    return @error_message = search.error_message unless search.valid?
 
-    @transportation_offices = TransportationOffice.by_distance_with_shipping_office(search.result).paginate(page: params[:page])
+    # return error message on invalid search
+    return error_message unless search.valid?
+
+    # return search results on valid search (GET)
+    return transportation_offices unless request.post?
+
+    # redirect on valid search (POST)
+    redirect_to offices_path(search.result)
   end
 
   private
 
+  def coordinates_search
+    CoordinatesSearch.new(params) if params[:latitude].present? && params[:longitude].present?
+  end
+
+  def error_message
+    @error_message ||= search.error_message
+  end
+
+  def installation_search
+    InstallationSearch.new(params) if params[:query].present?
+  end
+
   def search
-    @search ||= coordinates_search || zip_code_search || nil
+    @search ||= coordinates_search || zip_code_search || installation_search || nil
   end
 
-  def coordinates_search
-    CoordinatesSearch.new(params) if params[:coordinates].present?
+  def transportation_offices
+    @transportation_offices ||= TransportationOffice.by_distance_with_shipping_office(search.result).paginate(page: params[:page])
   end
 
   def zip_code_search
-    ZipCodeSearch.new(params) if params[:postal_code].present?
+    ZipCodeSearch.new(params) if params[:query].present? && /^\d{5}$/.match?(params[:query])
```

Comment:
bit of an edge case but it'd be nice to also support zip+4 codes since it's an easy change and wouldn't mess up any installation searches

Reply:
If we allow ZIP+4 searches, we would be implying that we can center the map on exactly that location, which would require a geocoding call to an external service. That's not out of the question, but it's not what we do now; we currently only use the external geocoding services to get the location of the offices during seeding IIRC, not on the live site.